File: LC 1202.cpp
class Solution {
public:
    string smallestStringWithSwaps(string s, vector<vector<int>>& pairs) {
        int m = s.length();
        vector<int>color(m,0);
        vector<vector<int>>g(m);
        for (const auto& v : pairs) {
            g[v[0]].push_back(v[1]);
            g[v[1]].push_back(v[0]);
        }

        auto sortString = [&] (vector<int>& v) {
            sort(v.begin(), v.end());
            vector<char>tmp;
            for (int x : v) {
                tmp.push_back(s[x]);
            }
            sort(tmp.begin(), tmp.end());
            for (size_t i = 0; i < v.size(); ++i) {
                s[v[i]] = tmp[i];
            }
        };

        function<void(int, vector<int>&)> dfs = [&] (int i, vector<int>& tmp) {
            if (color[i]) return;
            color[i] = 1;
            tmp.push_back(i);
            for (auto k : g[i]) {
                if (color[k] == 0)
                    dfs(k, tmp);
            } 
        };

        for (int i = 0; i < m; ++i) {
            if (color[i] == 0 && g[i].size() > 0) {
                vector<int>tmp;
                dfs(i, tmp);
                sortString(tmp);
            }
        }
        return s;
    }
};

File: 2000/LC 1202.go
func smallestStringWithSwaps(str string, pairs [][]int) string {
    n := len(str)
    s := []byte(str)
    a := []int{}
    g := make([][]int, n)
    for _, v := range pairs {
        a, b := v[0], v[1]
        g[a] = append(g[a], b)
        g[b] = append(g[b], a)
    }

    sortString := func() {
        tmp := []byte{}
        sort.Ints(a)
        for _, x := range a {
            tmp = append(tmp, s[x])
        }
        slices.Sort(tmp)
        for i, x := range a {
            s[x] = tmp[i]
        }
    }


    vis := make([]bool, n)
    var dfs func(int)
    dfs = func(i int) {
        if vis[i] {
            return 
        }
        vis[i] = true
        a = append(a, i)
        for _, x := range g[i] {
            dfs(x)
        }
    }

    for i := range s {
        if !vis[i] && len(g[i]) > 0 {
            a = nil
            dfs(i)
            sortString()
        }
    }

    return string(s)
}